### PR TITLE
Fix/homepageAnnouncementScroll

### PR DESCRIPTION
**Pull request by kishore03109**
## Problem

There are 2 main pending issues with Announcements. 
1. Announcments pinching up when updating them

https://github.com/isomerpages/isomercms-frontend/assets/42832651/6b9257e8-ecb7-4c0e-b1e3-77392b4831cc

2. Make 2 announcements, open the second announcement. Then, add a new announcement. Now, the opened announcement state is the second one, but since announcements are added from the top, this is a different announcement now.

https://github.com/isomerpages/isomercms-frontend/assets/42832651/d6cd253a-be7f-41e6-ac6c-c8a6798564a0

## Solution

for 1, I added a new array `announcementScrollRefs`. 

Working sol:

https://github.com/isomerpages/isomercms-frontend/assets/42832651/e13fc393-817a-48c4-af09-16b52244adf9

~for 2, i am blocked. I attempted until this [b0fa396](https://github.com/isomerpages/isomercms-frontend/pull/1530/commits/b0fa396498a367b8f67d4a2086a4cac38315ca77), and while it did fix the original error, it was visually buggy. Not sure if there is a more elegant way to fix this~

NOTE: Adding elements from the bottom as a workaround for now. 

https://github.com/isomerpages/isomercms-frontend/assets/42832651/d766126a-b549-486f-8195-1969f5dab53a

**Breaking Changes**

- [ ] Yes - this PR contains breaking changes
  - Details ...
- [X] No - this PR is backwards compatible with ALL of the following feature flags in this [doc](https://www.notion.so/opengov/Existing-feature-flags-518ad2cdc325420893a105e88c432be5)

### CODE REVIEW

**Comment by seaerchin** [comment]
double check - are we going to add from bottom now?

File: src/layouts/EditHomepage/EditHomepage.jsx [hunk 1]
```diff
@@ -723,6 +736,14 @@ const EditHomepage = ({ match }) => {
...
           )
           setHomepageState(updatedHomepageState)
+          setAnnouncementScrollRefs(
+            update(announcementScrollRefs, {
+              $push: [createRef()],
+            })
+          )
+          // We know this is the case since announcements are added from the top
+          const newAnnouncementIndex = 0
+          scrollTo(announcementScrollRefs[newAnnouncementIndex])
```
Comment (seaerchin) [inline src/layouts/EditHomepage/EditHomepage.jsx:746]: this might no longer be applicable here

File: src/layouts/EditHomepage/HomepagePreview.tsx [hunk 2]
```diff
@@ -175,6 +177,9 @@ export const HomepagePreview = ({
                 subtitle={section.announcements.subtitle}
                 announcementItems={section.announcements.announcement_items}
                 sectionIndex={sectionIndex}
+                announcementScrollRefs={
+                  announcementScrollRefs as Ref<HTMLDivElement>[]
+                }
```
Comment (seaerchin) [inline src/layouts/EditHomepage/HomepagePreview.tsx:182]: not an issue iwth this now but i just realised that instead of casting, we can just declare the type since we alr know it

File: src/layouts/components/Homepage/AnnouncementBody.tsx [hunk 3]
```diff
@@ -78,7 +78,22 @@ export const AnnouncementBody = ({
...
+                idx instanceof Array &&
+                idx.length > 0 &&
+                typeof idx[0] === "number"
+              ) {
+                onDisplay("announcement", idx[0])
+              } else {
+                onDisplay("announcement", -1)
+              }
+            }}
+          >
```
Comment (seaerchin) [inline src/layouts/components/Homepage/AnnouncementBody.tsx:96]: could we just add a comment here? alternatively, something we could do is change the API of `Accordion` (since we **do not** allow multiple) but it seems out of scope of this PR

File: src/templates/homepage/AnnouncementsSection.tsx [hunk 4]
```diff
@@ -16,15 +16,20 @@ type TemplateAnnouncementsSectionProps = Omit<
 
 export const TemplateAnnouncementsSection = forwardRef<
   HTMLDivElement,
-  TemplateAnnouncementsSectionProps
+  TemplateAnnouncementsSectionProps & {
+    announcementScrollRefs: Ref<HTMLDivElement>[]
+  }
```
Comment (seaerchin) [inline src/templates/homepage/AnnouncementsSection.tsx:21]: nit: prefer to have these on `TemplateAnnouncementsSectionProps`

File: src/layouts/EditHomepage/EditHomepage.jsx [hunk 5]
```diff
@@ -723,6 +736,14 @@ const EditHomepage = ({ match }) => {
             err
           )
           setHomepageState(updatedHomepageState)
+          setAnnouncementScrollRefs(
+            update(announcementScrollRefs, {
+              $push: [createRef()],
+            })
+          )
+          // We know this is the case since announcements are added from the top
+          const newAnnouncementIndex = 0
```
Comment (kishore03109) [inline src/layouts/EditHomepage/EditHomepage.jsx:745]: arh changing this now

File: src/layouts/EditHomepage/EditHomepage.jsx [hunk 6]
```diff
@@ -247,6 +249,7 @@ const EditHomepage = ({ match }) => {
         let highlightsErrors = []
         let announcementItemErrors = []
         const scrollRefs = []
+        const announcementScrollRefs = []
```
Comment (seaerchin) [inline src/layouts/EditHomepage/EditHomepage.jsx:252]: for clarity here, we have htis because we want to scroll to each individual announcement instead of the section as a whole right?

File: src/layouts/EditHomepage/EditHomepage.jsx (hunk 1 above)
Comment (kishore03109) [reply src/layouts/EditHomepage/EditHomepage.jsx]: fixed [ea67f34](https://github.com/isomerpages/isomercms-frontend/pull/1530/commits/ea67f34ea4ab01860270cb1da1787a1e75f5cef6)

File: src/layouts/EditHomepage/EditHomepage.jsx (hunk 6 above)
Comment (kishore03109) [reply src/layouts/EditHomepage/EditHomepage.jsx:252]: yes this is the case

File: src/layouts/EditHomepage/HomepagePreview.tsx (hunk 2 above)
Comment (kishore03109) [reply src/layouts/EditHomepage/HomepagePreview.tsx]: fair 2679320452e4a29dc282d5845fd0001eb0c4b5c2

File: src/layouts/components/Homepage/AnnouncementBody.tsx (hunk 3 above)
Comment (kishore03109) [reply src/layouts/components/Homepage/AnnouncementBody.tsx]: ya you are right this is abit confusing 8a7a0af058296742d3c7e252b404d4835afabadb

File: src/layouts/EditHomepage/HomepagePreview.tsx (hunk 2 above)
Comment (kishore03109) [reply src/layouts/EditHomepage/HomepagePreview.tsx]: 2679320452e4a29dc282d5845fd0001eb0c4b5c2

File: src/templates/homepage/AnnouncementsSection.tsx (hunk 4 above)
Comment (kishore03109) [reply src/templates/homepage/AnnouncementsSection.tsx]: 182daa87d8106f4c884cd674b09df274837508aa

File: src/templates/homepage/AnnouncementsSection.tsx [hunk 7]
```diff
@@ -24,7 +25,10 @@ export const TemplateAnnouncementsSection = forwardRef<
       subtitle,
       announcementItems,
       sectionIndex,
-    }: TemplateAnnouncementsSectionProps,
+      announcementScrollRefs,
+    }: TemplateAnnouncementsSectionProps & {
+      announcementScrollRefs: Ref<HTMLDivElement>[]
+    },
```
Comment (seaerchin) [inline src/templates/homepage/AnnouncementsSection.tsx:31]: supernit: this is extra alr since `TemplateAnnouncementsSectionProps` has the property.

**Review by seaerchin — APPROVED** [review]

File: src/templates/homepage/AnnouncementsSection.tsx (hunk 7 above)
Comment (kishore03109) [reply src/templates/homepage/AnnouncementsSection.tsx]: ps forgot to rm 54192b920cf2caf20804575260bfe5bfbe1178b1

**Review by seaerchin — APPROVED** [review]
rubber stamping
